File: rules/llm05-output-handling/python/exec-of-llm-output.py
import os
import sqlite3
import subprocess

from anthropic import Anthropic
from openai import OpenAI

openai_client = OpenAI()
claude = Anthropic()


def run_generated_code(prompt: str):
    resp = openai_client.chat.completions.create(
        model="gpt-4o", messages=[{"role": "user", "content": prompt}]
    )
    code = resp.choices[0].message.content
    # ruleid: exec-of-llm-output
    exec(code)


def eval_generated_expr(prompt: str):
    resp = claude.messages.create(
        model="claude-sonnet-5",
        max_tokens=1024,
        messages=[{"role": "user", "content": prompt}],
    )
    expr = resp.content[0].text
    # ruleid: exec-of-llm-output
    return eval(expr)


def run_generated_command(prompt: str):
    resp = openai_client.responses.create(model="gpt-4o", input=prompt)
    cmd = resp.output_text
    # ruleid: shell-from-llm-output
    os.system(cmd)


def run_generated_command_subprocess(prompt: str):
    resp = openai_client.chat.completions.create(
        model="gpt-4o", messages=[{"role": "user", "content": prompt}]
    )
    cmd = resp.choices[0].message.content
    # ruleid: shell-from-llm-output
    subprocess.run(cmd, shell=True)


def run_generated_sql(question: str, conn: sqlite3.Connection):
    resp = openai_client.chat.completions.create(
        model="gpt-4o",
        messages=[{"role": "user", "content": f"Write SQL for: {question}"}],
    )
    sql = resp.choices[0].message.content
    cur = conn.cursor()
    # ruleid: sql-from-llm-output
    cur.execute(sql)


def run_gemini_generated_code(genai_client, prompt: str):
    code = genai_client.models.generate_content(
        model="gemini-2.5-pro", contents=prompt
    ).text
    # ruleid: exec-of-llm-output
    exec(code)


def run_llamaindex_generated_sql(query_engine, conn, question: str):
    sql = query_engine.query(question).response
    cur = conn.cursor()
    # ruleid: sql-from-llm-output
    cur.execute(sql)


def safe_static_exec():
    # ok: exec-of-llm-output
    exec("print('hello')")


def safe_static_command():
    # ok: shell-from-llm-output
    subprocess.run(["ls", "-la"])


def safe_parameterized_sql(conn: sqlite3.Connection, user_id: int):
    cur = conn.cursor()
    # ok: sql-from-llm-output
    cur.execute("SELECT * FROM users WHERE id = ?", (user_id,))

File: rules/llm05-output-handling/typescript/exec-of-llm-output.ts
import Anthropic from "@anthropic-ai/sdk";
import { generateText } from "ai";
import { exec, execSync } from "child_process";
import OpenAI from "openai";

const openai = new OpenAI();
const claude = new Anthropic();

export async function runGeneratedCode(prompt: string) {
  const resp = await openai.chat.completions.create({
    model: "gpt-4o",
    messages: [{ role: "user", content: prompt }],
  });
  const code = resp.choices[0].message.content;
  // ruleid: eval-of-llm-output-js
  eval(code as string);
}

export async function evalGeneratedFn(prompt: string) {
  const resp = await claude.messages.create({
    model: "claude-sonnet-5",
    max_tokens: 1024,
    messages: [{ role: "user", content: prompt }],
  });
  const body = resp.content[0].text;
  // ruleid: eval-of-llm-output-js
  const fn = new Function(body);
  return fn();
}

export async function runGeneratedCommand(prompt: string) {
  const { text } = await generateText({ model: "openai/gpt-4o", prompt });
  // ruleid: shell-from-llm-output-js
  exec(text);
}

export async function runGeneratedCommandSync(prompt: string) {
  const resp = await openai.chat.completions.create({
    model: "gpt-4o",
    messages: [{ role: "user", content: prompt }],
  });
  const cmd = resp.choices[0].message.content;
  // ruleid: shell-from-llm-output-js
  execSync(cmd as string);
}

export function safeStaticEval() {
  // ok: eval-of-llm-output-js
  eval("1 + 1");
}

export function safeStaticCommand() {
  // ok: shell-from-llm-output-js
  execSync("ls -la");
}
